vissl/hooks/byol_hooks.py

```diff
@@ -0,0 +1,154 @@
+# Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved
+import math
+import logging
+
+import torch
+from classy_vision import tasks
+from classy_vision.generic.distributed_util import is_distributed_training_run
+from classy_vision.hooks.classy_hook import ClassyHook
+from vissl.models import build_model
+from vissl.utils.env import get_machine_local_and_dist_rank
+from vissl.utils.misc import concat_all_gather
+
+
+
+
+class BYOLHook(ClassyHook):
+    """
+    TODO: Update description
+
+    This hook corresponds to the loss proposed in the "Momentum Contrast
+    for Unsupervised Visual Representation Learning" paper, from Kaiming He et al.
+    See http://arxiv.org/abs/1911.05722 for details
+    and https://github.com/facebookresearch/moco for a reference implementation,
+    reused here.
+
+    Called after every forward pass to update the momentum encoder. At the beginning
+    of training i.e. after 1st forward call, the encoder is contructed and updated.
+    """
+
+    on_start = ClassyHook._noop
+    on_phase_start = ClassyHook._noop
+    on_loss_and_meter = ClassyHook._noop
+    on_backward = ClassyHook._noop
+    on_step = ClassyHook._noop
+    on_phase_end = ClassyHook._noop
+    on_end = ClassyHook._noop
+    on_update = ClassyHook._noop
+
+    def __init__(self, base_momentum: float, shuffle_batch: bool = True):
+        super().__init__()
+        self.base_momentum = base_momentum
+        self.is_distributed = False
+
+        self.momentum = None
+        self.inv_momentum = None
+        self.total_iters = None
+
+    @staticmethod
+    def cosine_decay(training_iter, max_iters, initial_value):
+        # TODO: Why do we need this min statement?
+        training_iter = min(training_iter, max_iters)
+        cosine_decay_value = 0.5 * (1 + math.cos(math.pi * training_iter / max_iters))
+        return initial_value * cosine_decay_value
+
+    @staticmethod
+    def target_ema(training_iter, base_ema, max_iters):
+        decay = BYOLHook.cosine_decay(training_iter, max_iters, 1.)
+        return 1. - (1. - base_ema) * decay
+
+    def _build_byol_target_network(self, task: tasks.ClassyTask) -> None:
+        """
+        Create the model replica called the target. This will slowly track
```

Comment:
Improve comment. Something like: "Target network is exponential moving average of  online network, ... "